ENH:MAINT:linalg:lu Cythonized and ndarray support added

ilayn opened the pull request:
Another PR for stacked versions of linalg funcs. This time it is the second and last leg of getting rid of FORTRAN module `flinalg` that was living inside `/scipy/linalg/src`. Old tests were a bit too trivial hence replaced with better ones.

It would be great if you can validate or offer new ideas for the newly introduced `p_indices` keyword. I can't think of a short and sweet alternative so went with the least confusing (to me) option. See [matlab handling it via `outputForm` parameter ](https://www.mathworks.com/help/matlab/ref/lu.html#mw_c7ac10f5-4b1d-4b24-aac5-cb08fc302896)(mathematica only returns indices anyways). I'll send a bike-shedding mail to the mailing list. 

I had to touch `matfuncs_expm.pyx.in` file to fix a generic problem I introduced when I cythonized it, but you can ignore it. It's a really big review (again, sigh... not fun authoring it either) but any feedback is welcome.



#### What does this implement/fix?
- Introduced a new keyword `p_indices` for returning permutation info as a vector instead of very costly matrices
- Removed the dependency to custom Fortran code in `/linalg/src/lu.f` (`_flinalg` module can be removed after this)
- Separated existing tests to their own file and added more
- Touched the documentation for cosmetics
- Performance is quite improved (3X-5X until roughly `n=100` then depends on LAPACK lib).
- Now supports arbitrary stacked arrays
- Also supports all numeric dtypes excluding float128, complex256.
- Handles all kinds of empty arrays
- Quick returns for `(1, 1)` and `(..., 1, 1)` shaped arrays

matusvalo commented on scipy/linalg/_cythonized_array_utils.pyx:
```diff
@@ -66,34 +66,23 @@ def find_det_from_lu(lapack_t[:, ::1] a):
-            c2 = c//2
-            swap_c_and_f_layout(a, b, r, c2, n);
-            swap_c_and_f_layout(a+(c2)*n, b+c2, r, c-c2, n)
+cdef inline void swap_c_and_f_layout(lapack_t *a, lapack_t *b, int r, int c) nogil:
```
```suggestion
cdef inline void swap_c_and_f_layout(lapack_t *a, lapack_t *b, int r, int c) noexcept nogil:
```

Just a humble suggestion: Cython 3 is going to change default handling of exceptions - see https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#exception-values-and-noexcept

Cython 0.29.31 and newer also support `noexcept` keyword as noop. Using `noexcept` keyword can simplify migration to Cython3.

There are already PRs related to this: https://github.com/scipy/scipy/pull/18266 and https://github.com/scipy/scipy/pull/18362

matusvalo commented on scipy/linalg/_decomp_lu_cython.pyx:
```diff
@@ -0,0 +1,160 @@
+cdef void lu_decompose(cnp.ndarray[lapack_t, ndim=2] a,
+                       cnp.ndarray[lapack_t, ndim=2] lu,
+                       int[::1] perm,
+                       bint permute_l):
```
```suggestion
                       bint permute_l) noexcept:
```

ilayn replied:
Just to make sure I understand correctly there is a possibility for this to raise exception below. Is this still needed here?

matusvalo replied:
> there is a possibility for this to raise exception below. Is this still needed here?

If there is possibility to have exception raised **and** you want to have correct code for both Cython 0.29.X and 3 you need to explicitly declare exception propagation using `except` keyword.

If you are writing code only for Cython 3 you can avoid specifying anything. I think it is worth to read Cython documentation about it: https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values It was pretty improved during recent days.

> **Warning**: using `void` return value can be dangerous when you forgot to provide `noexcept` mainly when `nogil` is also used - see following paragraf from cython documentation:
>
> There’s a known performance pitfall when combining nogil and except * @cython.exceptval(check=True). In this case Cython must always briefly re-acquire the GIL after a function call to check if an exception has been raised. This can commonly happen with a function returning nothing (C void). Simple workarounds are to mark the function as noexcept if you’re certain that exceptions cannot be thrown, or to change the return type to int and just let Cython use the return value as an error flag (by default, -1 triggers the exception check).

matusvalo replied:
> Cython definitely has a way of mentioning things in Reverse Polish notation. This is as confusing as the nogil keyword that with gil means gil is on, but nogil means it, maybe, might be off. Same with noexcept.

Yes GIL handling in cython can be easily confusing and I got lost in it multiple times.

> It is not that I have no exceptions possibly raised; but I don't know what noexcept means anymore. 

`noexcept` keyword means that you don't want to propagate the exception from function. It's the same as in Cython 0.29.X when you does not provide `except` keyword.

Consider following function compiled in Cython 0.29.X:

```cython
cdef int sum(int a, int b):
    raise ValueError
```

it is equal to following example (compiled in both Cython 0.29.X and Cython 3)

```cython
cdef int sum(int a, int b) noexcept:
    raise ValueError
```

If exception is raised in function body, it is not propagated, so if you write:

```python
try:
     sum(1, 2)
except ValueError:
     print("Got exception")
```

the `print("Got exception")` line is not executed for both examples.

I am sorry that I cannot explain it better :-(

ilayn replied:
Ah no you are too patient :) No problem at all. Sometimes my brain does not cooperate so it's all on me

ilayn commented:
The stats tests are having a hiccup again apparently.

ilayn commented:
@dschmitz89 are these vonmises failures look structural to you since you touched them most recently? 

https://github.com/scipy/scipy/actions/runs/4820761727/jobs/8585703287?pr=18358

mdhaber commented on scipy/linalg/tests/test_decomp.py:
```diff
@@ -913,144 +903,6 @@ def test_deprecation_results(self):
-        self.med = random((30, 40))
-        self.cmed = random((30, 40)) + 1.j * random((30, 40))
-
-    def _test_common(self, data):
```
It's not obvious that this was replaced (e.g. there is no longer a method with this name). Is this no longer needed, or was it replaced with the other new tests to `TestLU`? (Does it have to be replaced? It's easier to believe that things are safe if tests are not rewritten.)

ilayn replied:
All tests were combined for `lu`, `lu_factor`, and `lu_solve` and `test_common` was testing for `lu`. I took out the LU parts out and left the remaining ones in to not to cause any issues with those ones. At some point we can refactor those ones too then they will become the legacy tests.

mdhaber replied:
I'm not sure I understand.

> All tests were combined for `lu`, `lu_factor`, and `lu_solve`

`lu_solve` tests were separate from the `lu` and `lu_factor` tests before, and that has not changed.

>  I took out the LU parts out and left the remaining ones in to not to cause any issues with those ones.

Previously, `lu` and `lu_factor` were both tested in `TestLU`. By "left the remaining ones in", do you mean that you moved all the `lu_factor` tests into a new class `TestLUFactor`, and rewrote all the `lu` tests?

---

I'll try rephrasing my question:
Have all the `lu` tests been rewritten? If so, why? 
This test makes major changes to `lu`. If we want to ensure backward compatibility, we should *at least* make sure that all the old tests pass. Why not leave the old test alone, and add new ones as needed?

ilayn replied:
The old tests were, as you can follow from `lu_factor`, rather trivial small examples. So I tested with them anyways but they are far from testing anything substantial. The new tests spanning more sizes and cases. Old ones are just going to make noise with lots of manual array entries.

mdhaber replied:
As you wish, since I won't be merging. My hesitation is that I don't want to review the old tests to see that they are strictly being improved. So I suppose I will just run them locally to be sure that they still pass and otherwise ignore them, reviewing the new tests from scratch.

ilayn replied:
Sure no problem at all,

mdhaber commented on scipy/linalg/tests/test_decomp_lu.py:
```diff
@@ -0,0 +1,203 @@
+        assert_allclose(u, np.empty(shape=(1, 0, 0)))
+
+
+class TestLUFactor:
```
This looks equivalent (except without mixing in tests of `lu`) - just modernized.

ilayn replied:
Yes indeed, not modernized but maybe more readable conceptually.

mdhaber replied:
This comment is about `TestLUFactor` and `TestLUSolve`. By modernizing, I'm referring to changes like converting `assert_array_almost_equal` to the now-preferred `assert_allclose` and uses the singleton `RandomState` to an instance of the now-preferred (at least by some) `default_rng`.

ilayn replied:
Ah yes true, I consider that standard cleaning now. Didn't even occur to me :)

mdhaber reviewed (commented):
Like with `det`, I'll only be able to review the tests here. A few questions before I dig into the rewritten `TestLU`.

ilayn commented:
Other than the weird unicode error that I can't figure out, this is ready.

ilayn commented:
I managed to replicate the segfault in the failing test. The segfault doesn't happen when you just test the particular test file but segfaults, probably, runs out of memory when the full `test -s linalg` test is fired up. I'll change the LDL and cossin test sizes in a separate PR.

The linter issue is trivial.

h-vetinari commented on scipy/_lib/tests/test_public_api.py:
```diff
@@ -112,7 +112,6 @@ def test_dir_testing():
     'scipy.linalg.decomp_qr',
     'scipy.linalg.decomp_schur',
     'scipy.linalg.decomp_svd',
-    'scipy.linalg.flinalg',
```
Wouldn't we have to at least create a stub module that's still importable but raises a deprecation warning

ilayn replied:
I would like to risk that nobody on their right mind is importing det and lu from flinalg which emits deprecation warning already. 

There is always nonzero probability of course but I tend to choose not acting on it here. Purely my subjective opinion though. I can defer the removal to later version

h-vetinari replied:
Yeah, I had missed that these are emitting warnings already. The warning however does pretty clearly announce removal for 2.0:
```
>>> from scipy.linalg.flinalg import *
<stdin>:1: DeprecationWarning: The `scipy.linalg.flinalg` namespace is deprecated and will be removed in SciPy v2.0.0.
```
I'm not strictly against removing it, but it's a bit of inconsistent messaging (saying it'll be removed in 2.0, then removing earlier after all)

ilayn replied:
Somehow I managed to read it as 1.12.0... Dang.

ilayn replied:
@rgommers @tylerjereddy On one hand indeed this is breaking contract, on the other hand it is a very obscure module that only provides det and lu fortran api. Do you think it is safe to remove or should we leave it until SciPy 2.0 indeed?

rgommers replied:
I'd say that if it's just a matter of putting it back because you misread 2.0 as 1.12.0 and it's little effort, then put it back. But if it's more work to keep the deprecated and private API working because of other changes in this PR, then let's be pragmatic and get rid of it - this is not worth much effort either way.

ilayn replied:
OK added flinalg back. Thanks all.

h-vetinari commented on scipy/linalg/_cythonized_array_utils.pyx:
```diff
@@ -66,34 +66,26 @@ def find_det_from_lu(lapack_t[:, ::1] a):
+        for row in range(r):
+            bb[row] = aa[ith_row]
+            ith_row += c
+        aa += 1
+        bb += r
```
What's the reason for getting rid of the recursive descent here? Presumably this was added because it's faster to recurse into separate independent chunks (which don't hold the gil either, so can fully multithread) than going through one big loop?

I mean, I'm not against removing this (especially as performance seems to be better), but curious about this change.

ilayn replied:
It was me writing that code under the impression that recursion helped but I did some experiments and it in fact slows down. One of those theory does not reflect to practice on modern hardware thing. 

Mostly correcting my too-smart self from the past.

h-vetinari commented on scipy/linalg/_decomp_lu.py:
```diff
@@ -174,53 +182,166 @@ def lu(a, permute_l=False, overwrite_a=False, check_finite=True):
         Whether to check that the input matrix contains only finite numbers.
         Disabling may give a performance gain, but may result in problems
         (crashes, non-termination) if the inputs do contain infinities or NaNs.
+    p_indices : bool, optional
+        If ``True`` the permutation information is returned as row indices.
+        The default is ``False`` for backwards-compatibility reasons.
```
can we name this more closely to `permute_l`? something like `permutation_indices` / `perm_indices` / `permutation_idx`...

In general, I'd be happier if the replace-Fortran-with-Cython effort was separate from such API changes. I'm particularly not a fan of return signatures that change based on arguments. This is not the first and only place we have such issues, but then it would be nice to have a clear view of what API we want to achieve, and migrate there once, with a clear:
```
if legacy:
    # raises DeprecationWarning
    return the_old_way
else:
    return the_new_way  # possibly packaged in some smarter way, like LUResult
```

It would IMO be a pity to hold up the unequivocally beneficial fortran clean-up by the discussion that's (IMO) necessary for such API refactors.

ilayn replied:
There is a mailing list question about it that didn't receive much attention so I'm a bit undecided about the naming of `p_indices`. 

The return signature thing is very troubling me too but I don't know if we are ready to do it. It almost seems like a SciPy 2.0 type of clean up.

ilayn replied:
Revived the mailing list about it

h-vetinari commented on scipy/linalg/tests/test_decomp.py:
```diff
@@ -913,144 +903,6 @@ def test_deprecation_results(self):
         assert_allclose(v_dep, v)
 
 
-class TestLU:
-    def setup_method(self):
-        self.a = array([[1, 2, 3], [1, 2, 3], [2, 5, 6]])
-        self.ca = array([[1, 2, 3], [1, 2, 3], [2, 5j, 6]])
```
Could we not move these tests into a separate module in this PR? I'm happy to trust you that the tests didn't change and were just copy-and-pasted, but it's still extra review effort to verify that, especially if something were to go wrong afterwards. We could do this as a separate prep PR or afterwards?

ilayn replied:
True. The reason is that those are too simple tests. And very trivial to fit in the memory to test any array spillage etc.

That's why I replaced with more known examples further down the file. Tests are always nice to have but they should be a bit trying and not just perfectly singular arrays with integer entries. I can readd them anyways.

h-vetinari reviewed (commented):
Gave this a first pass so we can make some forward progress, though realistically, this and https://github.com/scipy/scipy/pull/18488 are both very big, and we probably have to pick and choose which one we want to focus on for 1.11 (or accept a lower level of review; I'm not suggesting that we _need_ to block)

ilayn commented:
Added the old tests, shrunk one of the large tests, renamed the new keyword to `matrix_p` to keep the `permute_l` rhyme. If tests pass, this is ready to go in.

adeak commented on scipy/linalg/_decomp_lu.py:
```diff
@@ -174,53 +182,166 @@ def lu(a, permute_l=False, overwrite_a=False, check_finite=True):
-        K = min(M, N)
-    u : (K, N) ndarray
-        Upper triangular or trapezoidal matrix
+    p : (..., M, M) ndarray
+        Permutation arrays or vectors depending on `matrix_p`
```
Just  to bikeshed a bit more:
1. Is it OK to call this `matrix_p` when now this can be a "stack of matrices" vs "stack of 1d arrays"? Probably yes, but we were already discussing the shade of the shed. But for a `(q, M, M)` shaped input and `matrix_p=False` we get shape `(q, M)` for `p`, right? Which would arguably "a matrix" as much as the `(M, M)` shaped original behaviour. Again, it's clear to me, and probably to most people, just want to make sure this is the least bad choice for the name.
2. Is it OK to keep the `(..., M, M)` shape in the docs here? Is there a way to spell this out correctly (or close enough) without building a 2x2 block of "if `thing` is `bool instane`" matrix?

ilayn replied:
Yes, the ndarray case is a generalization of 2D case hence storyline follows the 2D case and where available the sizes are replaced with it. I am practically following what NumPy does,  say with, https://numpy.org/doc/stable/reference/generated/numpy.linalg.qr.html . So the idea is that the returning thing is a lot of P matrices or a lot of P vectors in the ndarray case. 

But admittedly, story relies a lot on the legacy stuff. And to be honest, I don't know what the actual story line should be after looking at this line for a very long time.

adeak replied:
Just to be clear, my second point is that you have `(..., M, M)` whereas the linked NumPy `qr` has
```
‘raw’ : returns h, tau with dimensions (…, N, M), (…, K,)
```
which is exactly the kind of thing I'd expect for the new feature :)

adeak replied:
Oops, no I think I misread the above line. Let me reread...

adeak replied:
OK, [`numpy.linalg.svd`](https://numpy.org/doc/stable/reference/generated/numpy.linalg.svd.html)
```
    full_matrices : bool, optional
        If True (default), `u` and `vh` have the shapes ``(..., M, M)`` and
        ``(..., N, N)``, respectively.  Otherwise, the shapes are
        ``(..., M, K)`` and ``(..., K, N)``, respectively, where
        ``K = min(M, N)``.
```
So it spells out the two shapes with the flag, and then with the return value
```
    U : { (..., M, M), (..., M, K) } array
        Unitary array(s). The first ``a.ndim - 2`` dimensions have the same
        size as those of the input `a`. The size of the last two dimensions
        depends on the value of `full_matrices`. Only returned when
        `compute_uv` is True.
```
I think _this_ is indeed similar.

h-vetinari commented on scipy/linalg/_decomp_lu.py:
```diff
@@ -174,53 +182,166 @@ def lu(a, permute_l=False, overwrite_a=False, check_finite=True):
         Whether to check that the input matrix contains only finite numbers.
         Disabling may give a performance gain, but may result in problems
         (crashes, non-termination) if the inputs do contain infinities or NaNs.
+    matrix_p : bool, optional
+        If ``False`` the permutation information is returned as row indices.
+        The default is ``True`` for backwards-compatibility reasons.
```
From what I can tell based on the docstring example, we're not deprecating one variant for the other here? As in: both the vector of row indices as well as the full permutation matrix will be useful, depending on the usecase?

In that case, I don't think we should talk about backwards-compat at all - we're not planning to deprecate or remove anything (if both are useful). This would also allow us to avoid "leading" with `use False to do X`, which is the sort of inverted logic that's hard to remember. 

Also, I think we need to note the dependence on `permute_l` here, because this option makes no sense if `permute_l=True`. Perhaps such a combination should even error?

Perhaps something like the following?
```suggestion
    matrix_p : bool, optional
        If ``True`` (the default), return permutation information as a matrix;
        if ``False``, return permutation information as row indices.
        This option only takes effect if ``permute_l=False``.
```

ilayn replied:
reworded

h-vetinari commented on scipy/linalg/_decomp_lu_cython.pyi:
```diff
@@ -0,0 +1,6 @@
+
+def lu_decompose(a: NDArray[Any], lu: NDArray[Any], perm: NDArray[Any], permute_l: bool) -> None: ...
+
+def lu_dispatcher(a: NDArray[Any], lu: NDArray[Any], perm: NDArray[Any], permute_l: bool) -> None: ...
```
Missing newline

h-vetinari reviewed (commented):
Looking pretty good (though I didn't go over it with a fine-toothed comb)! One comment about the docstring.

Also while I appreciate that old tests were reinstated, the diff still makes it inscrutable what about the tests changed (or not). I'm not going to annoy you with this further, but FTR, the principle for such refactors should be: moving the tests from one module to another (here: the LU tests from `test_decomp.py` into `test_decomp_lu.py`) should come either _before_ or _after_ the PR that does the refactor, but not during.

ilayn commented:
It is not really annoyance, but it is a major task to define what the changes would look like from the start. Meaning that at some point you have to rewrite all commit history close to the merge, leaving old reviews stale. Here for example, I didn't care for the old tests but added after the reviews anyways. But I can't take them in as they were because they are from simpler times. 

I am in fact still learning how to do big PRs to juggle all the balls and while committing to commit hygiene. When things start to come together and then it comes to final tweaks, test tolerances here and there I wouldn't still know what a good commit workflow is while fixing all those. This is mainly because they fail only on CI and not locally. For example now I placed your comments in a review fixed up with an earlier to reduce the commit. Was it the right thing to do, should I have kept it for better review or is it ok to bury later changes in the earlier commits, I wouldn't know. It is as challenging as doing the actual PR for me.

h-vetinari commented:
> I am in fact still learning how to do big PRs to juggle all the balls and while committing to commit hygiene.

Yeah, it's non-trivial, and keeping a reasonable history (rather than just squashing everything into a mega-commit) can be a lot of work after the PR has gotten to a final state.

People have different (review) preferences too, so I won't impose mine, but I tend to:
* not care in tiny PRs (just adding incremental commits), because they'll be squashed
* rebase liberally in small-to-medium PRs where there's a small number of commits that should be reviewable stand-alone
* whatever works best for large PRs for _both_ the developer (myself) and the reviewers - either rebase often, or "finish development" and then do one big rebase (=rewrite git history) session.

Perhaps my requests are coming off as unrealistic because I've become used to interactive rebases, where - with a bit of experience - it's not an issue to create a commit that does a pure C&P of the tests into a separate module, and then rebase this PR on top of it. But I appreciate that it's a scary corner of git to get comfortable in.

h-vetinari commented:
However, one thing that _would_ be trivial even in this PR is simply not moving all those tests away from `test_decomp.py`. It's a completely cosmetic change (presumably to keep the test module size reasonable), and could come later too.

But again, not a demand, just explaining where I'm coming from when I ask these things. I try not to ask things that have a bad ratio of work-necessary-compared-to-achieved-gain.

ilayn commented:
Yes, `test_decomp.py` is nearing 3000 lines. That was our past mistake to let it grow as much. So it's a deferred pain in my opinion. Or conservation of dirt as control theory folks tend to say, you don't remove it just displace it until you run out of room. 

![image](https://github.com/scipy/scipy/assets/1303842/975b4339-1468-437b-a2cf-b5963c33e3d7)

h-vetinari commented:
> So it's a deferred pain in my opinion. 

To be clear, I was never suggesting _not_ to do it (in fact it could follow directly on the heels of this PR). I just expressed my preference to do such a change separately, because a big refactor is easier to review when the the test changes are small.

ilayn commented:
I wouldn't know how to do it both in place additions and then removing afterwards, it would be still as messy as is now to check the changes.

In fact that in my opinion makes things much harder. In this PR the old tests are kinda sorta ok to have, but for example with the other one in the sparse.linalg you need to change the tests to see if what you did actually worked, which turned out that it doesn't because all tests are practically tuned to a particular seed. 

So I understand the need, but in practice it is much much harder to preserve that order and clarity. Or at least I am not good at it probably. Because next thing you know you have touched 6 files and things are working; now do you have the patience to rewind back and distribute the changes to each commit  separately is the question. And for me the answer is typically no. And also I don't think that is feasible past some point

h-vetinari commented:
> I wouldn't know how to do it

If you're interested (either now or at some later point unrelated to this PR), I'm happy to show it, i.e. execute it, and provide the log of commands with notes. It would take me ~3min.

ilayn commented:
> If you're interested (either now or at some later point unrelated to this PR), I'm happy to show it,

I hope we are talking about the same thing but let's try anyways. 

In https://github.com/data-apis/array-api/issues/627 we received votes in favor of `p_indices` and possibly `p_as_indices` hence I am a bit leaning towards reverting the change but I really don't quite know, any after-thoughts from anyone?

ilayn commented:
I'd like to get this in but the name of the keyword issue is still remaining. On the other issue, @rgommers you hinted at `p_as_indices` support albeit with an thumbs up. I am also coming back to it after using it this week a bit more. @h-vetinari didn't like that initially and @adeak was considering the keyword name to be swapped to a name that is `True` by default. 

What do you folks think about this?

h-vetinari commented:
> On the other issue, @rgommers you hinted at `p_as_indices` support albeit with an thumbs up. I am also coming back to it after using it this week a bit more. @h-vetinari didn't like that

I don't care so much about the spelling of the keyword, I'll leave that to those more deeply involved in this. That said, my (weakly held) preferences are:
* something that rhymes with `permute_l`
* not indicating that one or the other is deprecated unless we have very good reason to. I'd just document the new option, because AFAICT both forms will be useful to someone (here's a [suggestion](https://github.com/scipy/scipy/pull/18358#discussion_r1201449717) based on my understanding)
* warn or error if an incompatible combination of `permute_l` and `p_as_indices` is used.

ilayn commented:
>  AFAICT both forms will be useful to someone 

The permutation info is definitely but representing it as a 2D array is not. So this is a historical oversight. We are at least trying to remedy the situation with providing a way to get the index version.

h-vetinari commented:
> The permutation info is definitely but representing it as a 2D array is not. So this is a historical oversight. We are at least trying to remedy the situation with providing a way to get the index version.

We have
```
    The decomposition satisfies::
        A = P @ L @ U
```
That's a nice property IMO, everything's just a matrix, everything can trivially be multiplied together. That doesn't work with indices[^1]. If someone picks up scipy with a basic linear algebra education but less programming experience, that simplicity seems like a benefit to me, despite the inefficient footprint of P.

[^1]: Yes, it works with `A = L[P, :] @ U`, but that's less obvious IMO

Do you feel the matrix return of `P` should be deprecated and eventually removed? I can see the benefit of a more efficient default (and then, the help-the-beginner argument doesn't really hold anymore anyway), so I'm not against it, I was just surprised because the case for removal of `P`-as-a-matrix didn't seem _that_ strong (though it's also not explicitly stated that eventual deprecation is the goal, only hinted at).

h-vetinari commented:
> Same with the changing number of arguments depending on keywords (matlab influence was particularly strong in the initial stages of SciPy).

I think that would be a great scipy 2.0 issue to ensure that we get rid of all of those.

ilayn commented:
OK since Tyler started to merge things, that means time to wrap up 😃  I'll revert to `p_indices` since folks really stuck to it then merge.

adeak commented:
> and @adeak was considering the keyword name to be swapped to a name that is `True` by default.

Just for completeness' sake: I only considered that because that seemed easier to name to me. I have no stance beyond that, any name works. Thanks.
